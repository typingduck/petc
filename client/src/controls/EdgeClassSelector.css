#petc-edge-select-control {
  background-color: var(--modal-bg);
  border-radius: var(--modal-border-radius);
  border: var(--modal-border);
  border-right-width: var(--modal-border-wide);
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  border-left: none;
  bottom: 50px;
  box-shadow: var(--modal-box-shadow);
  left: 0;
  padding: 0;
  position: absolute;
  top: 20px;
  z-index: var(--z-index-modal);
}

#petc-edge-select-control ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

#petc-edge-select-control li {
  align-items: center;
  border-bottom: 1px solid #555;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 2px;
  opacity: 0.8;
  padding: 10px;
}

#petc-edge-select-control li.selected {
  background-color: white;
}

.petc-edge-select-element {
  border: 1px solid rgba(0, 0, 0, 0);
  border-radius: 5px;
  height: 15px;
  position: relative;
  text-align: center;
  width: 80px;
}

.petc-edge-select-node-a {
  height: 10px;
  left: 0;
  position: absolute;
  width: 10px;
}

.petc-edge-select-node-b {
  height: 10px;
  position: absolute;
  right: 0;
  width: 10px;
}
