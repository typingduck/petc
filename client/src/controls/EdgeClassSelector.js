import React from 'react'

import './EdgeClassSelector.css'
import {jsPlumb} from 'jsplumb'

class EdgeClassSelector extends React.Component {
  render () {
    if (this.props.controls.isEdgeMode) {
      return <EdgeClassSelectorInternal {...this.props} />
    } else {
      return null
    }
  }
}

class EdgeClassSelectorInternal extends React.Component {
  constructor (props) {
    super(props)
    this.state = { jsPlmb: null }
    this.handleClassSelect = this.handleClassSelect.bind(this)
  }

  componentDidMount () {
    this.setState({
      jsPlmb: jsPlumb.getInstance({ Container: 'petc-edge-select-control' })
    })
  }

  classSelected (edgeClass) {
    return this.props.controls.selectedEdgeClass === edgeClass
  }

  handleClassSelect (edgeClass) {
    edgeClass = this.classSelected(edgeClass) ? null : edgeClass
    this.props.selectEdgeClass(edgeClass)
  }

  liClassName (className) {
    return this.classSelected(className) ? 'selected' : null
  }

  render () {
    const edgeClasses = (this.state.jsPlmb && Object.assign(
      { default: {} },
      (this.props.doc.style && this.props.doc.style.edges) || {}
    )) || {}
    return (
      <div id='petc-edge-select-control'>
        <ul>
          {Object.entries(edgeClasses).map(([className, style]) =>
            <li
              key={className}
              className={this.liClassName(className)}
              onClick={() => this.handleClassSelect(className)}
            >
              <EdgeClassElement
                edgeClassName={className}
                edgeStyle={style}
                jsPlmb={this.state.jsPlmb}
              />
              <div>{className}</div>
            </li>
          )}
        </ul>
      </div>
    )
  }
}

class EdgeClassElement extends React.Component {
  componentDidMount () {
    const edgeInfo = { source: this.endpointId('a'), target: this.endpointId('b') }
    Object.assign(edgeInfo, this.props.edgeStyle)
    this._connection = this.props.jsPlmb.connect(edgeInfo)
  }

  componentWillUnmount () {
    this.props.jsPlmb.deleteConnection(this._connection)
  }

  endpointId (endpoint) {
    return `petc-edge-select-element-${this.props.edgeClassName}-${endpoint}`
  }

  render () {
    return (
      <div className='petc-edge-select-element'>
        <div id={this.endpointId('a')} className='petc-edge-select-node-a' />
        <div id={this.endpointId('b')} className='petc-edge-select-node-b' />
      </div>
    )
  }
}

export default EdgeClassSelector
