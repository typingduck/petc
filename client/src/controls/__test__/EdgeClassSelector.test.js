/* global beforeEach expect it jest */
import React from 'react'
import Renderer from 'react-test-renderer'

import EdgeClassSelector from '../EdgeClassSelector'

jest.mock('uuid/v4', () => () => 'u-u-i-d')

const mockJsPlumb = {
  connect: jest.fn()
}

jest.mock('jsplumb', () => (
  { jsPlumb: { getInstance: () => mockJsPlumb } }
))

let props

beforeEach(() => {
  jest.clearAllMocks()
  props = {
    controls: {
      isNodeMode: false,
      isEdgeMode: true,
      selectedEdgeClass: null
    },
    doc: {
      style: {
        edges: {
          'default': {
            anchor: [ 'Left', 'Right' ]
          },
          'red-edge': {
            paintStyle: {
              stroke: 'red'
            }
          },
          'blue-edge': {
            paintStyle: {
              stroke: 'blue'
            }
          }
        }
      }
    },
    addEdge: jest.fn(),
    selectEdgeClass: jest.fn()
  }
})

it('hidden unless edge view mode', () => {
  const props = {
    controls: {
      isEdgeMode: false
    },
    doc: {}
  }
  const component = Renderer.create(<EdgeClassSelector {...props} />)
  expect(component.toJSON()).toBeNull()
})

it('visible when edge view mode', () => {
  const props = {
    controls: {
      isEdgeMode: true
    },
    doc: {}
  }
  const component = Renderer.create(<EdgeClassSelector {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('shows custom edge styles', () => {
  const component = Renderer.create(<EdgeClassSelector {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('highlights selected edge class', () => {
  props.controls.selectedEdgeClass = 'blue-edge'

  const component = Renderer.create(<EdgeClassSelector {...props} />)
  const ul = component.toJSON().children[0]

  const defaultEdge = edgeFromUl(ul, 0)
  expect(defaultEdge.props.className).toBeNull()

  const redEdge = edgeFromUl(ul, 1)
  expect(redEdge.props.className).toBeNull()

  const blueEdge = edgeFromUl(ul, 2)
  expect(blueEdge.props.className).toEqual(expect.stringContaining('selected'))
})

it('click to select edge', () => {
  const component = Renderer.create(<EdgeClassSelector {...props} />)
  const ul = component.toJSON().children[0]

  // drag a edge (only a short distance (so should be a click)
  const redEdge = edgeFromUl(ul, 1)
  simulateClickEdge(redEdge)
  expect(props.selectEdgeClass).toHaveBeenCalledWith('red-edge')
  props.controls.selectedEdgeClass = 'red-edge'

  const blueEdge = edgeFromUl(ul, 2)
  simulateClickEdge(blueEdge)
  expect(props.selectEdgeClass).toHaveBeenCalledWith('blue-edge')
  props.controls.selectedEdgeClass = 'blue-edge'

  // should deselect on second click
  simulateClickEdge(blueEdge)
  expect(props.selectEdgeClass).toHaveBeenCalledWith(null)
})

function edgeFromUl (ul, index) {
  return ul.children[index]
}

function simulateClickEdge (edgeLi) {
  edgeLi.props.onClick()
}
