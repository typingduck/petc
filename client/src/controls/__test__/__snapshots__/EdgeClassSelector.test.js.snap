// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows custom edge styles 1`] = `
<div
  id="petc-edge-select-control"
>
  <ul>
    <li
      className={null}
      onClick={[Function]}
    >
      <div
        className="petc-edge-select-element"
      >
        <div
          className="petc-edge-select-node-a"
          id="petc-edge-select-element-default-a"
        />
        <div
          className="petc-edge-select-node-b"
          id="petc-edge-select-element-default-b"
        />
      </div>
      <div>
        default
      </div>
    </li>
    <li
      className={null}
      onClick={[Function]}
    >
      <div
        className="petc-edge-select-element"
      >
        <div
          className="petc-edge-select-node-a"
          id="petc-edge-select-element-red-edge-a"
        />
        <div
          className="petc-edge-select-node-b"
          id="petc-edge-select-element-red-edge-b"
        />
      </div>
      <div>
        red-edge
      </div>
    </li>
    <li
      className={null}
      onClick={[Function]}
    >
      <div
        className="petc-edge-select-element"
      >
        <div
          className="petc-edge-select-node-a"
          id="petc-edge-select-element-blue-edge-a"
        />
        <div
          className="petc-edge-select-node-b"
          id="petc-edge-select-element-blue-edge-b"
        />
      </div>
      <div>
        blue-edge
      </div>
    </li>
  </ul>
</div>
`;

exports[`visible when edge view mode 1`] = `
<div
  id="petc-edge-select-control"
>
  <ul>
    <li
      className={null}
      onClick={[Function]}
    >
      <div
        className="petc-edge-select-element"
      >
        <div
          className="petc-edge-select-node-a"
          id="petc-edge-select-element-default-a"
        />
        <div
          className="petc-edge-select-node-b"
          id="petc-edge-select-element-default-b"
        />
      </div>
      <div>
        default
      </div>
    </li>
  </ul>
</div>
`;
